rfcs/0042-config-option.md

```diff
@@ -0,0 +1,301 @@
+
+## Occurences of problems
+
+- Because prometheus uses options to encode every possible setting, [#56017](https://github.com/NixOS/nixpkgs/pull/56017) is needed to allow users to set a part of the configuration that wasn't encoded yet.
+- Because strongswan-ctl uses options to encode its full configuration, changes like [#49197](https://github.com/NixOS/nixpkgs/pull/49197) are needed to update our options with upstream changes.
+- Pull requests like [#57036](https://github.com/NixOS/nixpkgs/pull/57036) or [#38324](https://github.com/NixOS/nixpkgs/pull/38324) are needed because users wish to have more configuration options than the ones provided.
+- [#58239](https://github.com/NixOS/nixpkgs/pull/58239), [#58181](https://github.com/NixOS/nixpkgs/pull/58181)
+
+## Previous discussions
+
+- https://github.com/NixOS/nixpkgs/pull/44923#issuecomment-412393196
+- https://github.com/NixOS/nixpkgs/pull/55957#issuecomment-464561483 -> https://github.com/NixOS/nixpkgs/pull/57716
+
+## Previous implementations
+
+This idea has been implemented already in some places:
+- [#45470](https://github.com/NixOS/nixpkgs/pull/45470)
+- [#52096](https://github.com/NixOS/nixpkgs/pull/52096)
+- [My Murmur module](https://github.com/Infinisil/system/blob/45c3ea36651a2f4328c8a7474148f1c5ecb18e0a/config/new-modules/murmur.nix)
+
+# Detailed design
+[design]: #detailed-design
+
+For specifying configuration files to programs in NixOS options, there should be a main single option called `config` which represents the configuration of the program as a Nix value, which can then be converted to the configuration file format the program expects. This may look as follows:
+
+```nix
+{ config, lib, ... }: with lib;
+let
+
+  cfg = config.services.foo;
+
+  configText = configGen.json cfg.config;
+
+in {
+
+  options.services.foo = {
+    enable = mkEnableOption "foo service";
+
+    config = mkOption {
+      type = lib.types.config.json;
+      default = {};
+      description = ''
+        Configuration for foo. Refer to <link xlink:href="https://example.com/docs/foo"/>
+        for documentation on the supported values.
+      '';
+    };
+  };
+  
+  config = mkIf cfg.enable {
+  
+    # Set minimal config to get service working by default
+    services.foo.config = {
+      # We don't use mkDefault, as this module requires this value in order to function
+      dataPath = "/var/lib/foo";
+      logLevel = mkDefault "WARN";
+      # Upstream default, needed for us to open the firewall
+      port = mkDefault 2546;
+    };
+  
+    environment.etc."foo.json".text = configText;
+    
+    networking.firewall.allowedTCPPorts = [ cfg.config.port ];
+    # ...
+  };
+}
+```
+
+This approach solves all* of the above mentioned problems. In addition we have the following properties that work with this approach out of the box:
```

Comment:
> This approach solves all* of the above mentioned problems.

I think this deserves some expanding. It is quite a bold claim that deserves some form of motivation. Granted, in most cases it is fairly obvious why this solves the problems listed in the Motivation section but it would still be beneficial to state it explicitly.

And in some cases it is not at all obvious, at least for me, and requires some thought. For example, the issue of "Upstream removes a setting (backwards incompatible)" seem to unresolved in the foo module above in case the `logLevel` option is removed or renamed upstream.

Reply:
Ah yeah, I'll expand this a bit, maybe mentioning that it solves the general [two points you mentioned](https://github.com/NixOS/rfcs/pull/42#discussion_r272852522).

And yeah, we are still tied to upstream for the settings we specify in the module itself, I guess that deserves a mention too.

Reply:
I consider this resolved, as I'm now mentioning that all of the problems are only solved for options we don't refer to in the module (https://github.com/NixOS/rfcs/pull/42/commits/e9082a92e0f97613cbdc9cb40ed0f6eb5a7ccd5e)